Handling of invalid ZIM files

veloman-yunkan opened the pull request:
Fixes #1128

When trying to display resources from an invalid ZIM file that fail to be retrieved due to problems with the ZIM data, now an HTML error page is returned instead. Note that the resource being accessed may in fact be an image, a CSS or a javascript file and the user may not see the error (though the appearance of the webpage trying to load and use that resource will be affected to various degrees).

To test the fix use the data from https://tmp.kiwix.org/corrupted_zim_crash/ and override the `library.xml` and `kiwix-desktop.session` files in your `~/.local/share/kiwix` directory with those found in the [attached ZIP archive](https://github.com/user-attachments/files/16597904/issue1128.zip). After `kiwix-desktop` is started it will contain two tabs:

1. Home page of the ZIM file - 4 books ("*Les misérables Tome I: Fantine*", "*Histoire de France 1440-1465 (Volume 7/19)*", "*Anciennes loix des François, conservées dans les coutumes angloises, recueillies par Littleton, Vol. I*", and "*Lexicon Latinum : $b Universae phraseologiae corpus congestum etc.*") will have their thumbnails replaced with a placeholder image because of the inability to load the image from the ZIM file (an HTML error page is instead served). As you can see it is a very weak indication that something is wrong with the ZIM file
2. The second tab refers to one of those images directly and the error is obvious

ShaopengLin reviewed (commented):
I still only got run_time error after following the instructions. Otherwise the change LGTM

kelson42 commented on resources/i18n/en.json:
```diff
@@ -174,6 +174,9 @@
     "zim-name": "Zim Name",
     "zim-path": "Zim File Path",
+    "bad-zim-file-error-page-title": "Invalid ZIM File",
+    "bad-zim-file-error-page-text": "ZIM file is invalid or corrupted.",
```
This should always be 'ZIM'

kelson42 reviewed (changes requested):
I have tested and it works fine. What I like is that to a large extend the app tries to deliver the content and fails only is a specific content is missing.

But we need:
* To put "ZIM" in strings (instead of "zim")
* The error page is not UTF8 charset declared, see:
* 
![image](https://github.com/user-attachments/assets/6e18fc46-baa7-45f0-9c57-fd623e517371)

mgautierfr commented on src/urlschemehandler.cpp:
```diff
@@ -216,6 +218,33 @@ UrlSchemeHandler::replyZimNotFoundPage(QWebEngineUrlRequestJob *request,
+    buffer->close();
+
+    connect(request, SIGNAL(destroyed()), buffer, SLOT(deleteLater()));
+    request->reply("text/html", buffer);
```
I wonder if we should not set a error code to something else than 200.
Request may be made by js script which will check the error code.
But I haven't found an API allowing to return a specific content and set error code in the same time.

veloman-yunkan replied:
I had the same difficulty while reviewing #1120. But since there we didn't have such a potential problem with accessing a resource via js, I [didn't mind](https://github.com/kiwix/kiwix-desktop/pull/1120#issuecomment-2151808886) about that scenario. Should we use [`QWebEngineUrlRequestJob::fail()`](https://doc.qt.io/qt-5/qwebengineurlrequestjob.html#fail) in this case (but then no special information about the error will be shown to the user)?

mgautierfr commented on src/urlschemehandler.cpp:
```diff
@@ -216,6 +218,33 @@ UrlSchemeHandler::replyZimNotFoundPage(QWebEngineUrlRequestJob *request,
+    const QString name = QString::fromStdString(book.getName());
+    const QString zimEntryPath = request->requestUrl().path();
+
+    QString contentHtml = "<section><div>"
+                          "<h1>" + gt("bad-zim-file-error-page-title") + "</h1>"
+                          "<p>"  + gt("bad-zim-file-error-page-text") + "</p>"
+                          "<p>"  + gt("zim-id")   + ": <b>" + zimId + "</b></p>"
+                          "<p>"  + gt("zim-name") + ": <b>" + name  + "</b></p>"
+                          "<p>"  + gt("zim-path") + ": <b>" + path  + "</b></p>"
+                          "<p>" +  gt("zim-entry-path") + ": <b>" + zimEntryPath + "</b></p>"
+                          "</div></section>";
```
Shouldn't we put all this text as a translatable template ?
RTL language may need to be formatted differently.

veloman-yunkan replied:
This problem was introduced in #1132, I only duplicated it. Maybe we can address it in a separate PR.

kelson42 replied:
I don't know about RTL, but the current way of doing is the one recommended by TW.

veloman-yunkan replied:
The concern is about "*key*: *value*" lines. How will they be displayed if *key* is in an RTL language/script and *value* is in a LTR script?

kelson42 replied:
Honestly pretty sure it will be all-right. Merging and we will fix if not, but this is anyway a general problem then.

veloman-yunkan commented:
Rebased on main and addressed review comments of @kelson42 and @ShaopengLin in new commits. @mgautierfr has identified a [good point](https://github.com/kiwix/kiwix-desktop/pull/1174#discussion_r1721720011) that has to be answered before we merge this PR or decide to change its implementation.

veloman-yunkan commented:
@kelson42 I think that you missed the [question](https://github.com/kiwix/kiwix-desktop/pull/1174#discussion_r1721720011) that had to be answered before merging the PR.

kelson42 commented:
@veloman-yunkan mist, was sure there was only one indeed.
